tools/update.php

```diff
@@ -0,0 +1,705 @@
+    //   [2] => /var/www/loris/SQL/Release_patches/18.0_To_19.0_upgrade.sql
+    //   [3] => /var/www/loris/SQL/Release_patches/19.0_To_19.1_upgrade.sql
+    // Range will run once when $start = $end. We want this in the case where
+    // there is no difference in major versions but there is a change in minor
+    // versions. The below loops will, in this case, match the minor release
+    // patches.
+    $start = $from_versions[MAJOR];
+    $end   = $start; // default case: there is a minor release difference
+    if ($diff_major > 0) {
+        $end = $to_versions[MAJOR];
+    }
+    $patch_dir           = $loris_root . 'SQL/Release_patches/';
+    $all_release_patches = glob($patch_dir . '*.sql');
+    $patches = [];
+    if ($diff_major > 0 || $diff_minor > 0) {
+        foreach (range($start, $end) as $v) {
+            foreach ($all_release_patches as $patch) {
+                // strpos must be 0 or else this will match w/ e.g. 16.0_To_17.0
+                // when the server is already on version 17
+                if (strpos(basename($patch), "$v") === 0) {
+                    $patches[] = $patch;
+                }
+            }
+        }
+    }
+    if (isDev()) {
+        echo "[+] Developer instance detected. Including developer patches..."
+            . PHP_EOL;
+
+        // Add all patches in Archive/$MAJOR.$MINOR. Everything other needed
+        // command will be in the Release patches which are been added above
+        $dev_patch_dir = $loris_root . 'SQL/Archive/' . $to_versions[MAJOR]
+            . '.' . $to_versions[MINOR] . '/';
+        $patches       = array_merge($patches, glob($dev_patch_dir . '*.sql'));
+    }
+    return $patches;
+}
+
+/**
+ * Apply a list of SQL patch files using exec.  Currently this function is only
+ * called with $report_only = true so patches are never applied. This can be
+ * changed in the future if this functionality is required.
+ *
+ * @param array      $patches     Patch files to be applied
+ * @param dictionary $db_config   DB config info. Needed for credentials
+ * @param bool       $report_only Whether to only display patches or to also
+ *                          apply them. Always true (for now).
+ *
+ * @return bool Always true (for now). False if patches fail to be applied.
+ */
+function applyPatches($patches, $db_config, $report_only = true) : bool
+{
+    // Iterate over all patches and source them into MySQL
+    $A = $db_config['database'];
+    $u = $db_config['username'];
+    $p = $db_config['password'];
+    $h = $db_config['host'];
+    if ($report_only === true) {
+        echo '[-] NOTE: Running in Report-Only mode. Patching commands will be '
+            . 'displayed but not executed.' . PHP_EOL;
+        sleep(1);
+    }
+    foreach ($patches as $patch) {
+        $cmd = "mysql -h $h -u $u -p -A $A < $patch";
+        if ($report_only === false) {
+            if (doExec($cmd) === false) {
+                break;
```

Comment:
so if I understand correctly, if a patch fails, doExec should spit out the error and this loop will break (and the script will terminate as this is the last action in main())

I think before breaking the rest of the commands should be output to the user along with a message indicating that the rest of the commands were not run. I figure if this didn't happen a user would just try to rerun this script which could make things worse.

Reply:
@ZainVirani that's a good point. I'll come back to this pending the conversation as to whether this script should even apply patche or if it should just print them.

Reply:
@ZainVirani I have added this functionality. Thanks for the suggestion!